### ISSUE TITLE
Keep getting shut-down Log Included

### ISSUE THREAD

**Issue by ShaunRoberts543** [rnilssoncx/homebridge-pico#24]
Hello, I keep getting the same response trying to set-up.  I had it working just fine, then it all went to hell.  I have reinstalled the plug-in.  Removed all Picos from Lutron and tried starting over.  It will throw this log after just setting up the bridge before even adding remotes.

TypeError: types[this.type] is not iterable
    at PicoRemote.getServices (/usr/local/lib/node_modules/homebridge-pico/accessory.js:71:29)
    at BridgeService.createHAPAccessory (/usr/local/lib/node_modules/homebridge/src/bridgeService.ts:456:41)
    at /usr/local/lib/node_modules/homebridge/src/bridgeService.ts:544:34
    at Array.forEach (<anonymous>)
    at /usr/local/lib/node_modules/homebridge/src/bridgeService.ts:534:21
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/util/once.ts:9:18
    at Pico.accessories (/usr/local/lib/node_modules/homebridge-pico/index.js:65:5)
    at /usr/local/lib/node_modules/homebridge/src/bridgeService.ts:529:24
    at new Promise (<anonymous>)
    at BridgeService.loadPlatformAccessories (/usr/local/lib/node_modules/homebridge/src/bridgeService.ts:523:12)

**Comment by rnilssoncx** [comment]
Can you post the pico section of your config file.

**Comment by ShaunRoberts543** [comment]
{
            "buslog": "full",
            "servers": [
                {
                    "switches": [
                        {
                            "pico": [
                                4
                            ],
                            "name": "PICO 15",
                            "type": "PJ2-4B-XXX-L01"
                        }
                    ],
                    "host": "192.168.50.218",
                    "port": 23
                }
            ],
            "longname": false,
            "platform": "Pico"
        }

**Comment by ShaunRoberts543** [comment]
I have 2 audio Picos set up natively in the Lutron app for my Sonos speaker.  Not sure if that would have any effect, just wanted to let it be known.

**Comment by ShaunRoberts543** [comment]
I reinstalled plug-in as noticed a possible fix on here.  Added a single remote and still get a sigterm but a different log:

TypeError: Cannot read property 'getCharacteristic' of undefined
    at PicoRemote.trigger (/usr/local/lib/node_modules/homebridge-pico/accessory.js:103:26)
    at Pico.clickHandler (/usr/local/lib/node_modules/homebridge-pico/index.js:80:45)
    at Click._finished (/usr/local/lib/node_modules/homebridge-pico/index.js:136:10)
    at listOnTimeout (internal/timers.js:555:17)
    at processTimers (internal/timers.js:498:7)

**Comment by rnilssoncx** [comment]
Can you provide all the Pico log entries before the crash - ideally from your several attempts.

**Comment by ShaunRoberts543** [comment]
28/05/2021, 10:18:44] [Pico] [192.168.50.218] Writing Keep Alive
[28/05/2021, 10:18:44] [Pico] [192.168.50.218] Bus Data: GNET> 
[28/05/2021, 10:18:51] [Hue] Family Room LED: set homekit status fault from 0 to 1
[28/05/2021, 10:19:01] [Hue] Family Room LED: set homekit status fault from 1 to 0
[28/05/2021, 10:19:10] [Pico] [192.168.50.218] Bus Data: ~DEVICE,4,11,3
[28/05/2021, 10:19:10] [Pico] [192.168.50.218] Device 4 Button 11 - Created tracker
[28/05/2021, 10:19:11] [Pico] [192.168.50.218] Bus Data: ~DEVICE,4,11,4
[28/05/2021, 10:19:11] TypeError: Cannot read property 'getCharacteristic' of undefined
    at PicoRemote.trigger (/usr/local/lib/node_modules/homebridge-pico/accessory.js:103:26)
    at Pico.clickHandler (/usr/local/lib/node_modules/homebridge-pico/index.js:80:45)
    at Click._finished (/usr/local/lib/node_modules/homebridge-pico/index.js:136:10)
    at listOnTimeout (internal/timers.js:555:17)
    at processTimers (internal/timers.js:498:7)
[28/05/2021, 10:19:11] Got SIGTERM, shutting down Homebridge...
[28/05/2021, 10:19:16] [Hue] goodbye
[28/05/2021, 10:19:16] [HB Supervisor] Homebridge Process Ended. Code: 143, Signal: null

**Comment by rnilssoncx** [comment]
There should be log messages from startup that show the pico setup - can you provide those as well.

**Comment by ShaunRoberts543** [comment]
[28/05/2021, 10:23:52] [Pico] [192.168.50.218] Writing Keep Alive
[28/05/2021, 10:23:52] [Pico] [192.168.50.218] Bus Data: GNET> 

[28/05/2021, 10:20:52] [Pico] [192.168.50.218] Caseta Pro Connected
Preparing Advertiser for 'Office TV D77F' using bonjour-hap backend!
Starting to advertise 'Office TV D77F' using bonjour-hap backend!

[28/05/2021, 10:20:53] [Pico] [192.168.50.218] Bus Data: login: 
[28/05/2021, 10:20:53] [Hue] Philips hue: Signify Netherlands B.V. BSB002 bridge v1944193080, api v1.44.0
[28/05/2021, 10:20:53] [Pico] [192.168.50.218] Bus Data: password: 

[28/05/2021, 10:20:53] [Pico] [192.168.50.218] Bus Data: GNET> 
[28/05/2021, 10:20:53] [Pico] [192.168.50.218] Logged in

[28/05/2021, 10:21:52] [Pico] [192.168.50.218] Writing Keep Alive
[28/05/2021, 10:21:52] [Pico] [192.168.50.218] Bus Data: GNET> 
[28/05/2021, 10:21:54] [Hue] Family Room LED: set homekit status fault from 1 to 0
[28/05/2021, 10:22:44] [Hue] Family Room LED: set homekit status fault from 0 to 1
[28/05/2021, 10:22:52] [Pico] [192.168.50.218] Writing Keep Alive
[28/05/2021, 10:22:52] [Pico] [192.168.50.218] Bus Data: GNET> 

[28/05/2021, 10:29:52] [Pico] [192.168.50.218] Writing Keep Alive
[28/05/2021, 10:30:52] [Pico] [192.168.50.218] Writing Keep Alive
[28/05/2021, 10:31:52] [Pico] [192.168.50.218] Writing Keep Alive
[28/05/2021, 10:32:52] [Pico] [192.168.50.218] Writing Keep Alive

**Comment by ShaunRoberts543** [comment]
Reset router and Lutron hub, restarted Homebridge and tried again.  Got the same when I pressed button on remote.

[28/05/2021, 10:35:43] [Pico] [192.168.50.218] Bus Data: ~DEVICE,4,11,3
[28/05/2021, 10:35:43] [Pico] [192.168.50.218] Device 4 Button 11 - Created tracker
[28/05/2021, 10:35:43] [Pico] [192.168.50.218] Bus Data: ~DEVICE,4,11,4
[28/05/2021, 10:35:43] TypeError: Cannot read property 'getCharacteristic' of undefined
    at PicoRemote.trigger (/usr/local/lib/node_modules/homebridge-pico/accessory.js:103:26)
    at Pico.clickHandler (/usr/local/lib/node_modules/homebridge-pico/index.js:80:45)
    at Click._finished (/usr/local/lib/node_modules/homebridge-pico/index.js:136:10)
    at listOnTimeout (internal/timers.js:555:17)
    at processTimers (internal/timers.js:498:7)
[28/05/2021, 10:35:43] Got SIGTERM, shutting down Homebridge...

**Comment by rnilssoncx** [comment]
The problem is that button "11" is not defined as part of the pico type you're selecting. 

Either the definition is wrong or you're selecting the wrong pico remote type. 

Can you provide me a link to a picture of the remote?

Also, if you change the id for the pico in the config so it won't register you'll be able to press each button and see the button number in the log. I need to know which button provides which number. 

Thanks.

**Comment by ShaunRoberts543** [comment]
![IMG_1367](https://user-images.githubusercontent.com/84977157/120001695-0be38780-bfa2-11eb-821e-55c144a7ea7d.jpeg)
![IMG_1368](https://user-images.githubusercontent.com/84977157/120001715-10a83b80-bfa2-11eb-986c-875e65fc21d9.jpeg)

4,8,3. 4,9,3. 4,10,3. 4,11,3.

**Comment by ShaunRoberts543** [comment]
So I switched the remote type to L31 as that is what the log is registering it as, a scene controller, and it appears to be working now.

**Comment by ShaunRoberts543** [comment]
Wait, maybe not.  It will not control anything.  I programmed a convert to shortcut on button 8...nothing.  I did a simple off command on button 9...nothing.

**Comment by rnilssoncx** [comment]
You have the right type selected.  Make sure you put your config back to watch for device 4.  I can see the challenge here - this particular PICO doesn't give you a full part number...  "scene" or "PJ2-4B-XXX-L31".

If the logs show the Pico trackers being created and the presses being sent to homekit you're good on the homebridge side.

I will put in some additional checks this weekend to help make sure the sort of problem you have had here with the config will be less painful.  The plugin needs to grow up a bit and not expect everything will be perfect.

Glad to help if your issues persist, but I believe we're past any config problems.

**Comment by ShaunRoberts543** [comment]
It is an awesome plug-in.  When I had it working it was super responsive.  I am actually going around replacing other buttons with picos for that reason.  Hopefully I can get this to register with HomeKit to actually control things.  I restarted my router as previously stated and it sometimes takes forever to populate everything, that may be the cause currently.  I have like 130 devices on my system, so it drags a bit during restarts.

**Comment by searls** [comment]
I'm seeing the same error. I just installed the plugin and have not added anything yet aside from pointing the config to my pro hub.

Here's my config:

```json
        {
            "buslog": "monitor",
            "servers": [
                {
                    "switches": [
                        {
                            "pico": []
                        }
                    ],
                    "host": "192.168.1.200",
                    "port": 23
                }
            ],
            "longname": false,
            "platform": "Pico"
        }
```

And here's the relevant section of the log (the homebridge server itself is continuously rebooting itself so it's not clear which comes first)
```
[25/07/2021, 14:03:37] TypeError: types[this.type] is not iterable
    at PicoRemote.getServices (/usr/local/lib/node_modules/homebridge-pico/accessory.js:71:29)
    at BridgeService.createHAPAccessory (/usr/local/lib/node_modules/homebridge/src/bridgeService.ts:456:41)
    at /usr/local/lib/node_modules/homebridge/src/bridgeService.ts:544:34
    at Array.forEach (<anonymous>)
    at /usr/local/lib/node_modules/homebridge/src/bridgeService.ts:534:21
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/util/once.ts:9:18
    at Pico.accessories (/usr/local/lib/node_modules/homebridge-pico/index.js:65:5)
    at /usr/local/lib/node_modules/homebridge/src/bridgeService.ts:529:24
    at new Promise (<anonymous>)
    at BridgeService.loadPlatformAccessories (/usr/local/lib/node_modules/homebridge/src/bridgeService.ts:523:12)
[25/07/2021, 14:03:37] [Pico] [192.168.1.200] Caseta Pro Connected
[25/07/2021, 14:03:37] Got SIGTERM, shutting down Homebridge...
[25/07/2021, 14:03:37] [Pico] [192.168.1.200] Bus Data: login: 
[25/07/2021, 14:03:37] [Pico] [192.168.1.200] Bus Data: password: 
```

**Comment by rnilssoncx** [comment]
A quick workaround for your issue is to define a pico remote - just make up a number and give it any valid type. I'm working on an update that will survive missing config items, etc.

**Comment by searls** [comment]
@rnilssoncx right on, thanks for the pointer. Really appreciate that you're working on and sharing this with the public. I know it's thankless work

**Comment by rnilssoncx** [comment]
This should now be fixed - currently in 1.1.1-beta.0.  I'll release the updated version in a few days if there are no issues.
